utils/comparable_bytes.cc

```diff
@@ -23,6 +25,120 @@ static void convert_signed_fixed_length_integer(managed_bytes_view& src, bytes_o
     out.write<T>(read_simple<T>(src) ^ sign_mask);
 }
 
+/**
+ * Variable-length encoding for signed long data type (int64_t).
+ * The first bit of the coding stores the inverted sign bit followed by as many matching bits as
+ * there are additional bytes in the encoding, followed by the two's complement of the number.
+ *
+ * (i.e) <n bits with inverted sign bit><2's complement of the number>
+ *       where n = number of bytes in encoding and n >= 1
+ *
+ * Because of the inverted sign bit, negative numbers compare smaller than positives, and because the length
+ * bits match the sign, longer positive numbers compare greater and longer negative ones compare smaller.
+ *
+ * Examples:
+ *      0              encodes as           80
+ *      1              encodes as           81
+ *     -1              encodes as           7F
+ *     63              encodes as           BF
+ *     64              encodes as           C040
+ *    -64              encodes as           40
+ *    -65              encodes as           3FBF
+ *   2^20-1            encodes as           EFFFFF
+ *   2^20              encodes as           F0100000
+ *  -2^20              encodes as           100000
+ *   2^64-1            encodes as           FFFFFFFFFFFFFFFFFF
+ *  -2^64              encodes as           000000000000000000
+ *
+ * The encoded value will have a maximum of 9 bytes including the length bits.
+ * As the number of bytes is specified in bits 2-9, no value is a prefix of another.
+ */
+static void encode_signed_long_type(managed_bytes_view src, bytes_ostream& out) {
+    // Create a byte filled with the sign bit
+    uint8_t curr_byte = read_simple<uint8_t>(src);
+    const uint8_t sign_only_byte = (int8_t(curr_byte) >> 7);
+
+    // Discard the leading sign-only bytes
+    while (curr_byte == sign_only_byte && !src.empty()) {
+        curr_byte = read_simple<uint8_t>(src);
+    }
+
+    // The value to be encoded is in curr_byte and src
+    const size_t leading_sign_bits_in_curr_byte = sign_only_byte ? std::countl_one(curr_byte) : std::countl_zero(curr_byte);
+    size_t num_bits = (8 - leading_sign_bits_in_curr_byte) + src.size_bytes() * 8;
+    // Calculate the number of length bits (inverted sign bits in the prefix)
+    // that indicate the total size of the encoded format in bytes.
+    // 0-6 bits => 1 byte; 7-13 => 2 bytes; etc to 56-63 => 9 bytes.
+    // (63 bits will need 10 bytes, but this encoding does it with 9)
+    size_t num_length_bits_to_write = std::min(size_t(9), num_bits / 7 + 1);
+
+    if (num_length_bits_to_write >= 8) {
+        // Write one byte full of length bits (i.e. inverted sign bits)
+        out.write<uint8_t>(~sign_only_byte);
```

Comment:
Use a temporary byte array (of size 9) to avoid engaging the bytes_ostream machinery for every byte. It is heavy since it needs to check for out-of-space each time.